### PR TITLE
Add logic for key validation in report subcommand

### PR DESCRIPTION
**Pull request by subham-deepsource**
If the wrong value is passed to `--key` in `deepsource report`, it would send it to the backend and return a cryptic error message.

This ensures that you pass a supported language to `--key`.

### CODE REVIEW

**Comment by subham-deepsource** [comment]
To make the tests pass, I think we have to update the DSN. @siddhant-deepsource Right?

**Comment by siddhant-deepsource** [comment]
Its the issue with using the test coverage github action. Ref: https://discuss.deepsource.io/t/breaking-deepsource-test-coverage-github-action/507

File: command/report/report.go [hunk 1]
```diff
@@ -52,16 +52,16 @@ func NewCmdReport() *cobra.Command {
 	}
 
 	// --repo, -r flag
-	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "The analyzer shortcode whose test coverage is being reported")
+	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "name of the analyzer (example: test-coverage)")
```
Comment (anto-deepsource) [inline command/report/report.go:55]: The descriptions for `analyzer` and `key` are confusing, users might confuse `analyzer` as the language for which they are uploading the artifact. IMO, we can be more descriptive.

I'll think about and suggest something but in the meantime, if you get a better description, you can update it.

**Review by anto-deepsource — CHANGES_REQUESTED** [review]

File: command/report/report.go (hunk 1 above)
Comment (tushar-deepsource) [inline command/report/report.go:55]:
```suggestion
	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "name of the analyzer to report the artifact to (example: test-coverage)")
```

File: command/report/report.go [hunk 2]
```diff
@@ -52,16 +52,16 @@ func NewCmdReport() *cobra.Command {
...
+	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "name of the analyzer to report the artifact to (example: test-coverage)")
 
-	cmd.Flags().StringVar(&opts.Key, "key", "", "The artifact being reported.tcv for test-coverage")
+	cmd.Flags().StringVar(&opts.Key, "key", "", "shortcode of the analyzer (example: go)")
 
-	cmd.Flags().StringVar(&opts.ValueFile, "value-file", "", "Path to the value file")
+	cmd.Flags().StringVar(&opts.Value, "value", "", "value of the artifact")
 
-	cmd.Flags().StringVar(&opts.Value, "value", "", "Value of the artifact")
+	cmd.Flags().StringVar(&opts.ValueFile, "value-file", "", "path of the artifact value file")
```
Comment (tushar-deepsource) [inline command/report/report.go:61]:
```suggestion
	cmd.Flags().StringVar(&opts.ValueFile, "value-file", "", "path to the artifact value file")
```

**Comment by tushar-deepsource** [comment]
CI is fixed now.

File: command/report/report.go [hunk 3]
```diff
@@ -52,16 +52,16 @@ func NewCmdReport() *cobra.Command {
 	}
 
 	// --repo, -r flag
-	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "The analyzer shortcode whose test coverage is being reported")
+	cmd.Flags().StringVar(&opts.Analyzer, "analyzer", "", "name of the analyzer to report the artifact to (example: test-coverage)")
 
-	cmd.Flags().StringVar(&opts.Key, "key", "", "The artifact being reported.tcv for test-coverage")
+	cmd.Flags().StringVar(&opts.Key, "key", "", "shortcode of the analyzer (example: go)")
```
Comment (tushar-deepsource) [inline command/report/report.go:57]:
```suggestion
	cmd.Flags().StringVar(&opts.Key, "key", "", "shortcode of the language (example: go)")
```

**Review by anto-deepsource — DISMISSED** [review]

File: command/report/tests/report_workflow_test.go [hunk 4]
```diff
@@ -50,9 +52,15 @@ func graphQLAPIMock(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 	w.Header().Set("Content-Type", "application/json")
 
-	if string(requestBodyData) == string(req) {
+	if want, got := string(requestBodyData), string(req); want == got {
 		w.Write([]byte(successResponseBodyData))
 	} else {
+		if eq := deep.Equal(want, got); len(eq) != 0 {
```
Comment (sourya-deepsource) [inline command/report/tests/report_workflow_test.go:58]: `go-cmp` is already being installed. You can use `cmp.DeepEqual`.

File: command/report/tests/report_workflow_test.go [hunk 5]
```diff
@@ -50,9 +52,15 @@ func graphQLAPIMock(w http.ResponseWriter, r *http.Request) {
...
+			log.Println("Mismatch found:")
+			for _, e := range eq {
+				log.Printf("diff: %v\n", e)
+			}
```
Comment (sourya-deepsource) [inline command/report/tests/report_workflow_test.go:62]: You can use `cmp.Diff` instead.

**Review by sourya-deepsource — CHANGES_REQUESTED** [review]

File: command/report/report.go [hunk 6]
```diff
@@ -97,6 +97,30 @@ func (opts *ReportOptions) Run() int {
...
+		"java":       true,
+		"scala":      true,
+		"php":        true,
+		"csharp":     true,
+		"cxx":        true,
+		"rust":       true,
+	}
+
+	if !supportedKey[reportCommandKey] {
+		err = fmt.Errorf("DeepSource | Error | Invalid Key: %s", reportCommandKey)
```
[the remark this answers is not in the document]
Comment (subham-deepsource) [reply command/report/report.go]: Now it looks like:
```
DeepSource | Error | Invalid Key: abc (Supported Keys: [python javascript java csharp rust go ruby scala php cxx])
```

**Review by sourya-deepsource — APPROVED** [review]

**Pull request merged**